gpui: Provide workaround for AMD Linux driver bug

not-my-profile opened the pull request:
Closes #26143

There apparently is some amdgpu/radv bug that rendering with
multisample anti-aliasing (MSAA) results in a crash when the bounds
of a triangle list exceed 1024px, which in Zed happens with the default
buffer font size when you select a line with more than 144 characters.

This crash has been reported as #26143.

This commit introduces a workaround: you can set the ZED_PATH_SAMPLE_COUNT=0
environment variable to disable MSAA and the error message we print
when a GPU crash is encountered with radv now suggests trying this environment
variable as a workaround and links the respective issue.

Sidenote: MSAA was introduced in f08b1d78ec1655a56b07c5efcd74d225d2a4b7d5
so you didn't run into this driver bug with versions < 0.173.8.

Release Notes:

- Added a workaround for an AMD Linux driver bug that causes Zed to crash when selecting long lines.

jansol commented:
How is multisampling related to text selection? :confused:

not-my-profile commented:
Good question. I don't know. But with PATH_SAMPLE_COUNT = 4 the bug is 100% reproducible for me, that is zed always crashes when I select a line with more than 114 characters and when I set it to 1 then it no longer happens. I did note that the amount of characters in a line that trigger it depends somehow on the buffer font size but it's not simply the total width of the line in pixels that also changes depending on the font size.

jansol commented:
According to https://vulkan.gpuinfo.org/displayreport.php?id=37319#properties_core_10 https://vulkan.gpuinfo.org/displayreport.php?id=37075#properties_core_10 both of your GPUs support up to 8 samples, so this is highly suspicious. (enter "SampleCount" in the search field)

not-my-profile commented:
https://vulkan.lunarg.com/issue/view/5eef839ce0d2a30448a06559 seems relevant. I can look into it later today.

mikayla-maki commented:
Given the uncertainty in this bug fix, I'm going to close this PR for now.

not-my-profile commented:
@mikayla-maki I'm certain that it fixes the crash and I said that I would look into it (to find out what exactly is going on and if there's a better fix for it). I'd understand converting this to a draft but just outright closing this feels very unwelcoming.

mikayla-maki commented:
@not-my-profile Fair, I'm trying to bias more towards close-and-reopen so we don't have stale drafts lying around. But this PR is fresh enough that I think I erred on the wrong side on this one :)

not-my-profile commented:
I figured out that this is somehow caused by the rounded corners of the highlighted range line (in combination with MSAA >1). If I set the `corner_radius` to `0` it no longer happens. Rounded corners of course change the tessellation but the vertices for a path that works and one that crashes the GPU only look subtly different (they even have the same vertices count). I'll investigate more when I have time later this week.

@mikayla-maki could we do a hot fix to disable the rounded corners of text selections on Linux for AMD GPUs (detecting them via `window.gpu_specs()`) until we've figured out the root cause?

not-my-profile commented:
Ok, it's triggered by one triangle being pushed to `gpui::scene::Path` having a side that's more than 1035px long.
I tried splitting up the triangle with multiple `push_triangle` calls but that didn't work.

not-my-profile committed:
gpui: Provide workaround for AMD Linux driver bug

There apparently is some amdgpu/radv bug that rendering with
multisample anti-aliasing (MSAA) results in a crash when the bounds
of a triangle list exceed 1024px, which in Zed happens with the default
buffer font size when you select a line with more than 144 characters.

This crash has been reported as #26143.

This commit introduces a workaround: you can set the ZED_PATH_SAMPLE_COUNT=0
environment variable to disable MSAA and the error message we print
when a GPU crash is encountered with radv now suggests trying this environment
variable as a workaround and links the respective issue.

Sidenote: MSAA was introduced in f08b1d78ec1655a56b07c5efcd74d225d2a4b7d5
so you didn't run into this driver bug with versions < 0.173.8.

mikayla-maki reviewed (changes requested):
Thanks for switching this to an environment variable! Could you also add some docs about this workaround to https://github.com/zed-industries/zed/blob/066761902ae80390dca87a1a400281519b955ca8/docs/src/linux.md?plain=1#L92 ?

huacnlee commented:
> How is multisampling related to text selection? :confused: 

Because text selection used `Path` to build polygon as the selection background, and in Zed there is only Path used MSAA.

This is looks like this GPU driver bug? They are AMD graphics cards is not too old.

@not-my-profile Have you test for this on Windows? Maybe this is Linux driver bug.

not-my-profile commented:
I agree that this looks like a driver bug. I unfortunately don't have Windows to test this. I build a small [reproduction app](https://github.com/not-my-profile/debug-radv-msaa) and am hoping to gain some insight by users trying it out in #26143.

Interestingly enough the boundaries that trigger it seem to be changing. For me most often it's 1025px but I also had it happen that it only crashed on 1026px and now somebody reported that even 1024px is crashing for them.

not-my-profile commented:
nitpick: The pre-release notes say the issue is fixed. It's not, I merely added a workaround.

mmmarkk01 commented:
It is definitely not fixed, also setting the environment variables does not work on my end.

jansol commented:
Have you tried reporting this to the mesa bug tracker to get the driver fixed?

not-my-profile commented:
@mmmarkk01 ZED_PATH_SAMPLE_COUNT=0 was introduced in v0.179.2. Are you on zed >=0.179.2?

@jansol It's not clear yet to me if the bug is in amdgpu or radv. In #26143 a user reported that upgrading their kernel from 6.9.3 to 6.12.10 fixed the issue for them, which would hint to it being in amdgpu and already fixed. I haven't had time to investigate further yet. In any case I think we should keep the discussion to the issue.

mmmarkk01 commented:
> @mmmarkk01 ZED_PATH_SAMPLE_COUNT=0 was introduced in v0.179.2. Are you on zed >=0.179.2?
> 
> 
> 
> yes i am, it keeps crashing on lines longer than 110'ish.
